python/cudf_polars/cudf_polars/experimental/groupby.py

```diff
@@ -0,0 +1,304 @@
+
+    # Construct final selection expressions
+    col_expr: Col | BinOp | UnaryFunction
+    dtype = schema[name]
+    complex_expr = complex_expr_map.get(name, None)
+    if complex_expr is None:
+        col_expr = Col(dtype, name)
+    elif "mean" in complex_expr:
+        mean_cols = complex_expr["mean"]
+        col_expr = BinOp(
+            dtype,
+            plc.binaryop.BinaryOperator.DIV,
+            Col(dtype, mean_cols["sum"]),
+            Col(dtype, mean_cols["count"]),
+        )
+    if unary_op:
+        col_expr = UnaryFunction(
+            dtype,
+            unary_op["name"],
+            unary_op["options"],
+            col_expr,
+        )
+    selection_exprs.append(NamedExpr(name, col_expr))
+
+    return piecewise_exprs, reduction_exprs, selection_exprs
+
+
+@lower_ir_node.register(GroupBy)
+def _(
+    ir: GroupBy, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # Extract child partitioning
+    child, partition_info = rec(ir.children[0])
+
+    # Handle single-partition case
+    if partition_info[child].count == 1:
+        single_part_node = ir.reconstruct([child])
+        partition_info[single_part_node] = partition_info[child]
+        return single_part_node, partition_info
+
+    # Check group-by keys
+    if not all(expr.is_pointwise for expr in traversal([e.value for e in ir.keys])):
+        raise NotImplementedError(
+            f"GroupBy does not support multiple partitions for keys:\n{ir.keys}"
+        )  # pragma: no cover
+
+    # Check if we are dealing with any high-cardinality columns
+    post_aggregation_count = 1  # Default tree reduction
+    groupby_key_columns = [ne.name for ne in ir.keys]
+    cardinality_factor = {
+        c: min(f, 1.0)
+        for c, f in ir.config_options.get("executor_options", {})
+        .get("cardinality_factor", {})
+        .items()
+        if c in groupby_key_columns
+    }
+    if cardinality_factor:
+        # The `cardinality_factor` dictionary can be used
+        # to specify a mapping between column names and
+        # cardinality "factors". Each factor estimates the
+        # fractional number of unique values in the column.
+        # Each value should be in the range (0, 1].
+        child_count = partition_info[child].count
+        post_aggregation_count = max(
+            int(max(cardinality_factor.values()) * child_count),
+            1,
+        )
+
+    # Decompose the aggregation requests into three distinct phases
+    piecewise_exprs, reduction_exprs, selection_exprs = (
+        list(itertools.chain.from_iterable(x))
+        for x in zip(
+            *map(
+                partial(decompose_groupby_reduction, ir.schema),
+                ir.agg_requests,
+            ),
+            strict=False,
+        )
+    )
+
+    # Partition-wise groupby operation
+    gb_pwise = GroupBy(
+        ir.schema,
+        ir.keys,
+        piecewise_exprs,
+        ir.maintain_order,
```

Comment:
I don't think we can support `maintain_order == True` (at least easily). So perhaps we should raise in that case.

Reply:
I think we support it for the tree reduction, but not for a shuffle-based reduction, right? The tree-reduction tasks should be ordered appropriately.

Reply:
Ah true.